### PR TITLE
Homepage headers to compose

### PR DESCRIPTION
**Pull request by sarah541**
Fixes https://github.com/mozilla-mobile/fenix/issues/23970
Fixes https://github.com/mozilla-mobile/fenix/issues/23969
Fixes https://github.com/mozilla-mobile/fenix/issues/23968

### Pull Request checklist

- [ ] **Tests**: This PR includes thorough tests or an explanation of why it does not
- [ ] **Screenshots**: This PR includes screenshots or GIFs of the changes made or an explanation of why it does not
- [ ] **Accessibility**: The code in this PR follows [accessibility best practices](https://github.com/mozilla-mobile/shared-docs/blob/master/android/accessibility_guide.md) or does not include any user facing features. In addition, it includes a screenshot of a successful [accessibility scan](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor&hl=en_US) to ensure no new defects are added to the product.

### To download an APK when reviewing a PR:
1. click on Show All Checks,
2. click Details next to "Taskcluster (pull_request)" after it appears and then finishes with a green checkmark,
3. click on the "Fenix - assemble" task, then click "Run Artifacts".
4. the APK links should be on the left side of the screen, named for each CPU architecture

### CODE REVIEW

**Comment by sarah541** [comment]
Disabling 2 UI tests which will be addressed in #24315

File: app/src/main/java/org/mozilla/fenix/compose/home/HomeSectionHeader.kt [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+/**
+ * Homepage header.
+ *
+ * @param text The header string.
+ * @param (optional) onShowAllClick Action to take when show all is clicked.
+ */
+@Composable
+fun HomeSectionHeader(
+    text: String,
+    description: String,
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/compose/home/HomeSectionHeader.kt:34]:
```suggestion
    contentDescription: String,
```
Can you also add an @param

File: app/src/main/java/org/mozilla/fenix/compose/home/HomeSectionHeader.kt [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+/**
+ * Homepage header.
+ *
+ * @param text The header string.
+ * @param (optional) onShowAllClick Action to take when show all is clicked.
+ */
+@Composable
+fun HomeSectionHeader(
+    text: String,
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/compose/home/HomeSectionHeader.kt:33]: Maybe rename this to be `headerText`

File: app/src/main/java/org/mozilla/fenix/home/recentbookmarks/view/RecentBookmarksHeaderViewHolder.kt [hunk 3]
```diff
@@ -18,26 +29,52 @@ import org.mozilla.fenix.utils.view.ViewHolder
...
-            dismissSearchDialogIfDisplayed()
-            interactor.onShowAllBookmarksClicked()
-        }
+        val horizontalPadding =
+            composeView.resources.getDimensionPixelSize(R.dimen.home_item_horizontal_margin)
+        composeView.setPadding(horizontalPadding, 0, horizontalPadding, 0)
     }
 
-    private fun dismissSearchDialogIfDisplayed() {
+    private fun dismissSearchDialogIfDisplayedAndShowAllClicked() {
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/recentbookmarks/view/RecentBookmarksHeaderViewHolder.kt:43]: I don't think we should combine the 2. Let's just keep the call similar to before.
```
onShowAllClick = {
            dismissSearchDialogIfDisplayed()
            interactor.onShowAllBookmarksClicked()
}
```

File: app/src/main/java/org/mozilla/fenix/home/recentvisits/view/RecentVisitsHeaderViewHolder.kt [hunk 4]
```diff
@@ -17,18 +28,44 @@ import org.mozilla.fenix.utils.view.ViewHolder
...
+}
+
+@Composable
+@Preview
+private fun RecentVisitsHeaderPreview() {
+    FirefoxTheme {
+        HomeSectionHeader(
+            stringResource(R.string.history_metadata_header_2),
+            stringResource(R.string.past_explorations_show_all_content_description_2)
+        )
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/recentvisits/view/RecentVisitsHeaderViewHolder.kt:69]: Let's just go with a single @Preview in `HomeSectionHeader.kt`

File: app/src/main/res/layout/recent_visits_header.xml [hunk 5]
```diff
@@ -1,43 +0,0 @@
...
-        app:layout_constraintEnd_toStartOf="@id/show_all_button"
-        app:layout_constraintTop_toTopOf="parent" />
-
-    <TextView
-        android:id="@+id/show_all_button"
-        style="@style/Button14TextStyle"
-        android:layout_width="wrap_content"
-        android:layout_height="wrap_content"
-        android:background="?android:attr/selectableItemBackground"
-        android:contentDescription="@string/past_explorations_show_all_content_description_2"
```
Comment (gabrielluong) [inline app/src/main/res/layout/recent_visits_header.xml:33]: I think this string can also be removed?

File: app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlAdapter.kt [hunk 6]
```diff
@@ -250,6 +250,21 @@ class SessionControlAdapter(
                 interactor = interactor,
                 metrics = components.analytics.metrics
             )
+            RecentVisitsHeaderViewHolder.LAYOUT_ID -> return RecentVisitsHeaderViewHolder(
+                composeView = ComposeView(parent.context),
+                viewLifecycleOwner,
+                interactor
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlAdapter.kt:256]: Can you add the named parameter for these?

**Review by gabrielluong — CHANGES_REQUESTED** [review]
Some minor changes, but I think we are close

File: app/src/main/java/org/mozilla/fenix/compose/home/HomeSectionHeader.kt (hunk 1 above)
Comment (gabrielluong) [reply app/src/main/java/org/mozilla/fenix/compose/home/HomeSectionHeader.kt:44]: Were we unable to call this `contentDescription`?

File: app/src/main/java/org/mozilla/fenix/compose/home/HomeSectionHeader.kt [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+    }
+}
+
+/**
+ * Homepage header.
+ *
+ * @param headerText The header string.
+ * @param description The description for click action
+ * @param navController to dismiss search dialog
+ * @param (optional) onShowAllClick Action to take when show all is clicked.
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/compose/home/HomeSectionHeader.kt:39]:
```suggestion
 * @param onShowAllButtonClick Invoked when "Show all" button is clicked.
```

File: app/src/main/java/org/mozilla/fenix/compose/home/HomeSectionHeader.kt [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+ */
+@Composable
+fun HomeSectionHeader(
+    headerText: String,
+    description: String,
+    navController: NavController?,
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/compose/home/HomeSectionHeader.kt:45]: Don't need this

File: app/src/main/java/org/mozilla/fenix/compose/home/HomeSectionHeader.kt [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+            text = AnnotatedString(text = stringResource(id = R.string.recent_tabs_show_all)),
+            modifier = Modifier.semantics {
+                contentDescription = description
+            },
+            style = TextStyle(
+                color = FirefoxTheme.colors.textAccent,
+                fontSize = 14.sp
+            ),
+            onClick = {
+                dismissSearchDialogIfDisplayed(navController)
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/compose/home/HomeSectionHeader.kt:69]: Don't need this

File: app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabsHeaderViewHolder.kt [hunk 10]
```diff
@@ -36,7 +45,25 @@ class RecentTabsHeaderViewHolder(
...
+    override fun Content() {
+        Column {
+            Spacer(modifier = Modifier.height(40.dp))
+
+            HomeSectionHeader(
+                headerText = stringResource(R.string.recent_tabs_header),
+                description = stringResource(id = R.string.recent_tabs_show_all_content_description_2),
+                onShowAllClick = {
+                    dismissSearchDialogIfDisplayed()
+                    interactor::onRecentTabShowAllClicked
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/recenttabs/view/RecentTabsHeaderViewHolder.kt:58]:
```suggestion
                    interactor.onRecentTabShowAllClicked()
```

File: app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlAdapter.kt [hunk 11]
```diff
@@ -284,12 +299,6 @@ class SessionControlAdapter(
                 view
             )
             ExperimentDefaultBrowserCardViewHolder.LAYOUT_ID -> ExperimentDefaultBrowserCardViewHolder(view, interactor)
-            RecentTabsHeaderViewHolder.LAYOUT_ID -> RecentTabsHeaderViewHolder(view, interactor)
-            RecentBookmarksHeaderViewHolder.LAYOUT_ID -> RecentBookmarksHeaderViewHolder(view, interactor)
-            RecentVisitsHeaderViewHolder.LAYOUT_ID -> RecentVisitsHeaderViewHolder(
-                view,
```
Comment (gabrielluong) [inline app/src/main/java/org/mozilla/fenix/home/sessioncontrol/SessionControlAdapter.kt:290]: Can you also add these in `onViewRecycled` below?

**Review by gbrownmozilla — APPROVED** [review]

**Pull request closed without merging**